Support data preprocessing in Spark framework

jackjlli opened the pull request:
## Description
Currently the data preprocessing Hadoop job is located in pinot-hadoop module. And there is no data preprocessing Spark job. In order to reused some of the logic of data preprocessing (e.g. support data in AVRO and ORC format) for both MR and Spark framework, the common code is moved from pinot-hadoop to pinot-ingestion-common module.

This PR:
* adds data preprocessing support for Spark.
* refactors some code from pinot-hadoop to pinot-ingestion-common so that some part of the code can be reused for both MR and Spark job.
* both AVRO and ORC formats are supported in both MR and Spark job.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
<!-- If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release. -->

<!-- If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text.
-->
## Documentation
<!-- If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document
-->

Commit:
Support data preprocessing in Spark framework

kishoreg commented:
is this needed only for v0_deprecated/spark? Can you please add more to the description on what is the change and why it's needed.

jackjlli commented:
> is this needed only for v0_deprecated/spark? Can you please add more to the description on what is the change and why it's needed.

I've updated the description of the PR. Basically this PR is to add the data preprocessing Spark job in pinot. And since some logic can be reused for both MR and Spark jobs, I refactored some code from pinot-hadoop module to pinot-ingestion-common module. At LinkedIn this data preprocessing job has already been used along with the PinotBuildAndPushJob from v0_deprecated module.

klsince commented on pinot-plugins/pinot-batch-ingestion/v0_deprecated/pinot-spark/src/main/java/org/apache/pinot/spark/jobs/preprocess/SparkDataPreprocessingHelper.java:
```diff
@@ -0,0 +1,153 @@
+    // Repartition and sort within partitions.
+    Comparator<Object> comparator = new SparkDataPreprocessingComparator();
+    JavaPairRDD<Object, Row> partitionedSortedPairRDD =
+        pairRDD.repartitionAndSortWithinPartitions(sparkPartitioner, comparator);
```
sometimes use coalesce() (then sortWithinPartitions()) can be more efficient than repartition(). is it feasible to allow coalesce()?

klsince commented on pinot-plugins/pinot-batch-ingestion/v0_deprecated/pinot-spark/src/main/java/org/apache/pinot/spark/jobs/preprocess/SparkDataPreprocessingPartitioner.java:
```diff
@@ -0,0 +1,56 @@
+
+  @Override
+  public int getPartition(Object key) {
+    SparkDataPreprocessingJobKey jobKey = (SparkDataPreprocessingJobKey) key;
```
why not just call generatePartitionId() in this method?

jackjlli replied:
The key of the KV pair is not only used for partitioning but also used for sorting purpose. It'd be good to generate the partition index in order to minimize the footprint of the key, since any data type can be the partition column (e.g. string).

xiangfu0 commented:
Shall we consider use pinot-ingestion-spark module?

jackjlli commented:
> Shall we consider use pinot-ingestion-spark module?

Good question! I also thought about putting it to pinot-ingestion-spark module, while this data preprocessing job for Spark uses common code which is also used for the preprocessing MR job in pinot-hadoop module. That's why I put it under the same 'v0_deprecated/' directory for now. The core logic should be simple and can be easily put to pinot-ingestion-spark module later on.

jackjlli commented on pinot-plugins/pinot-batch-ingestion/v0_deprecated/pinot-ingestion-common/src/main/java/org/apache/pinot/ingestion/jobs/SegmentPreprocessingJob.java:
```diff
@@ -90,4 +119,131 @@ protected boolean isDataFile(String fileName) {
+      }
+    }
+    if (!sortingColumns.isEmpty()) {
+      Preconditions.checkArgument(sortingColumns.size() == 1, "There should be at most 1 sorted column in the table.");
```
If sortingColumns.size() is 0 which also qualify for `size() <= 1`, `sortingColumns.get(0);` will return NullPointerException. Thus, a if block is still needed anyway.

klsince replied:
ack, a if-check would be needed for `.get(0)` to be safe. I was mainly concerning about the error msg `at most 1` and the `== 1` check is a bit inconsistent. I was suggesting sth like L191 and L192 below.

jackjlli replied:
Updated it to `size() <= 1`.

jackjlli commented on pinot-plugins/pinot-batch-ingestion/v0_deprecated/pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java:
```diff
@@ -140,13 +140,16 @@ public void setup(Context context)
 
     // Set up segment name generator
-    String segmentNameGeneratorType = _jobConf.get(JobConfigConstants.SEGMENT_NAME_GENERATOR_TYPE, JobConfigConstants.DEFAULT_SEGMENT_NAME_GENERATOR);
+    String segmentNameGeneratorType =
```
Basically the reformatting is just for adding data preprocessing support on spark framework. And all the renaming changes are under the same parent module. And I've already address with the latest master, so no big reformatting changes here. So I think it should be fine to just keep it in one PR.

xiangfu0 commented:
Closing this PR as stale because it has had no activity for more than one year (last updated before 2025-06-18). If this is still relevant, please reopen it or open a fresh PR against current master. Thanks for the contribution.
